File: frontend/src/types/learningPath.ts
/**
 * Types for Learning Path features
 */

export enum DifficultyLevel {
  BEGINNER = 'beginner',
  INTERMEDIATE = 'intermediate',
  ADVANCED = 'advanced',
}

export enum ProgressStatus {
  NOT_STARTED = 'not_started',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
}

export enum PathItemType {
  COURSE = 'course',
  ASSIGNMENT = 'assignment',
  QUIZ = 'quiz',
  RESOURCE = 'resource',
}

export interface LearningPathItem {
  id: number;
  learning_path_id: number;
  item_type: PathItemType;
  item_id: number;
  title: string;
  description?: string;
  order_index: number;
  is_required: boolean;
  estimated_hours?: number;
  prerequisites?: number[];
  created_at: string;
  updated_at: string;
}

export interface LearningPath {
  id: number;
  title: string;
  description?: string;
  difficulty_level: DifficultyLevel;
  estimated_hours?: number;
  icon?: string;
  color?: string;
  is_active: boolean;
  created_by_id?: number;
  created_at: string;
  updated_at: string;
  tags: string[];
}

export interface LearningPathDetail extends LearningPath {
  items: LearningPathItem[];
  total_items: number;
  required_items: number;
}

export interface UserPathItemProgress {
  id: number;
  user_id: number;
  path_item_id: number;
  status: ProgressStatus;
  score?: number;
  started_at?: string;
  completed_at?: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface UserLearningProgress {
  id: number;
  user_id: number;
  learning_path_id: number;
  status: ProgressStatus;
  progress_percentage: number;
  started_at?: string;
  completed_at?: string;
  last_accessed_at?: string;
  created_at: string;
  updated_at: string;
}

export interface LearningPathItemWithProgress extends LearningPathItem {
  user_progress?: UserPathItemProgress;
  is_locked: boolean;
}

export interface LearningPathWithProgress extends LearningPathDetail {
  user_progress?: UserLearningProgress;
  items_with_progress: LearningPathItemWithProgress[];
}

export interface LearningPathRecommendation {
  learning_path: LearningPath;
  recommendation_score: number;
  recommendation_reason: string;
  matching_tags: string[];
  user_progress?: UserLearningProgress;
}

export interface RecommendationsResponse {
  recommendations: LearningPathRecommendation[];
  total: number;
  user_completed_paths: number;
  user_in_progress_paths: number;
}

export interface EnrollmentRequest {
  learning_path_id: number;
}

export interface EnrollmentResponse {
  success: boolean;
  message: string;
  user_progress: UserLearningProgress;
}

export interface UserLearningStats {
  total_paths_enrolled: number;
  total_paths_completed: number;
  total_paths_in_progress: number;
  total_items_completed: number;
  total_learning_hours: number;
  completion_rate: number;
  recent_activity: Array<{
    path_id: number;
    path_title: string;
    status: ProgressStatus;
    progress: number;
    last_accessed?: string;
  }>;
}

export interface LearningPathCreateRequest {
  title: string;
  description?: string;
  difficulty_level?: DifficultyLevel;
  estimated_hours?: number;
  icon?: string;
  color?: string;
  is_active?: boolean;
  tags?: string[];
  items?: Array<{
    item_type: PathItemType;
    item_id: number;
    title: string;
    description?: string;
    order_index: number;
    is_required?: boolean;
    estimated_hours?: number;
    prerequisites?: number[];
  }>;
}

export interface UserPathItemProgressUpdate {
  status: ProgressStatus;
  score?: number;
  notes?: string;
}

File: backend/app/schemas/learning_path.py
"""
Pydantic schemas for Learning Path API
"""
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from enum import Enum


class DifficultyLevel(str, Enum):
    """Difficulty levels"""
    BEGINNER = "beginner"
    INTERMEDIATE = "intermediate"
    ADVANCED = "advanced"


class ProgressStatus(str, Enum):
    """Progress status"""
    NOT_STARTED = "not_started"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"


class PathItemType(str, Enum):
    """Types of learning path items"""
    COURSE = "course"
    ASSIGNMENT = "assignment"
    QUIZ = "quiz"
    RESOURCE = "resource"


# Learning Path Item Schemas
class LearningPathItemBase(BaseModel):
    """Base schema for learning path item"""
    item_type: PathItemType
    item_id: int
    title: str
    description: Optional[str] = None
    order_index: int
    is_required: bool = True
    estimated_hours: Optional[int] = None
    prerequisites: Optional[List[int]] = None  # List of item IDs that must be completed first


class LearningPathItemCreate(LearningPathItemBase):
    """Schema for creating a learning path item"""
    pass


class LearningPathItemUpdate(BaseModel):
    """Schema for updating a learning path item"""
    title: Optional[str] = None
    description: Optional[str] = None
    order_index: Optional[int] = None
    is_required: Optional[bool] = None
    estimated_hours: Optional[int] = None
    prerequisites: Optional[List[int]] = None


class LearningPathItemResponse(LearningPathItemBase):
    """Schema for learning path item response"""
    id: int
    learning_path_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Learning Path Schemas
class LearningPathBase(BaseModel):
    """Base schema for learning path"""
    title: str
    description: Optional[str] = None
    difficulty_level: DifficultyLevel = DifficultyLevel.BEGINNER
    estimated_hours: Optional[int] = None
    icon: Optional[str] = None
    color: Optional[str] = None
    is_active: bool = True


class LearningPathCreate(LearningPathBase):
    """Schema for creating a learning path"""
    tags: Optional[List[str]] = None
    items: Optional[List[LearningPathItemCreate]] = None


class LearningPathUpdate(BaseModel):
    """Schema for updating a learning path"""
    title: Optional[str] = None
    description: Optional[str] = None
    difficulty_level: Optional[DifficultyLevel] = None
    estimated_hours: Optional[int] = None
    icon: Optional[str] = None
    color: Optional[str] = None
    is_active: Optional[bool] = None
    tags: Optional[List[str]] = None


class LearningPathResponse(LearningPathBase):
    """Schema for learning path response"""
    id: int
    created_by_id: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    tags: List[str] = []

    class Config:
        from_attributes = True


class LearningPathDetailResponse(LearningPathResponse):
    """Schema for detailed learning path response with items"""
    items: List[LearningPathItemResponse] = []
    total_items: int = 0
    required_items: int = 0


# User Progress Schemas
class UserPathItemProgressBase(BaseModel):
    """Base schema for user path item progress"""
    status: ProgressStatus = ProgressStatus.NOT_STARTED
    score: Optional[float] = None
    notes: Optional[str] = None


class UserPathItemProgressUpdate(UserPathItemProgressBase):
    """Schema for updating user path item progress"""
    pass


class UserPathItemProgressResponse(UserPathItemProgressBase):
    """Schema for user path item progress response"""
    id: int
    user_id: int
    path_item_id: int
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class UserLearningProgressBase(BaseModel):
    """Base schema for user learning progress"""
    status: ProgressStatus = ProgressStatus.NOT_STARTED
    progress_percentage: float = 0.0


class UserLearningProgressResponse(UserLearningProgressBase):
    """Schema for user learning progress response"""
    id: int
    user_id: int
    learning_path_id: int
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    last_accessed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Combined schemas for detailed views
class LearningPathItemWithProgress(LearningPathItemResponse):
    """Learning path item with user progress"""
    user_progress: Optional[UserPathItemProgressResponse] = None
    is_locked: bool = False  # True if prerequisites not met


class LearningPathWithProgress(LearningPathDetailResponse):
    """Learning path with user progress"""
    user_progress: Optional[UserLearningProgressResponse] = None
    items_with_progress: List[LearningPathItemWithProgress] = []


# Recommendation schemas
class LearningPathRecommendation(BaseModel):
    """Schema for learning path recommendation"""
    learning_path: LearningPathResponse
    recommendation_score: float = Field(..., ge=0, le=100)
    recommendation_reason: str
    matching_tags: List[str] = []
    user_progress: Optional[UserLearningProgressResponse] = None


class RecommendationsResponse(BaseModel):
    """Schema for recommendations response"""
    recommendations: List[LearningPathRecommendation]
    total: int
    user_completed_paths: int
    user_in_progress_paths: int


# Enrollment schema
class EnrollmentRequest(BaseModel):
    """Schema for enrolling in a learning path"""
    learning_path_id: int


class EnrollmentResponse(BaseModel):
    """Schema for enrollment response"""
    success: bool
    message: str
    user_progress: UserLearningProgressResponse

    class Config:
        from_attributes = True


# Statistics schemas
class LearningPathStats(BaseModel):
    """Statistics for a learning path"""
    learning_path_id: int
    total_enrolled: int
    total_completed: int
    total_in_progress: int
    completion_rate: float
    average_completion_time_hours: Optional[float] = None


class UserLearningStats(BaseModel):
    """Statistics for a user's learning"""
    total_paths_enrolled: int
    total_paths_completed: int
    total_paths_in_progress: int
    total_items_completed: int
    total_learning_hours: float
    completion_rate: float
    recent_activity: List[Dict[str, Any]] = []
